### ISSUE TITLE
pgr_drivingDistance returns no results when array contains disconnected item

### ISSUE THREAD

**Issue by AlexBulankou** [pgRouting/pgrouting#519]
pgRouting extension version=2.1 (built from https://github.com/pgRouting/pgrouting/archive/pgrouting-2.1.0.zip)

I've been using pgr_drivingDistance to retrieve the list of shortest pairwise distances from each node in my select query to each node in the array that I'm passing as second parameter. I know that [documentation ](http://docs.pgrouting.org/2.0/en/src/driving_distance/doc/dd_driving_distance.html)doesn't really mention that array can be passed as second parameter, but so far it worked - please let me know if this is no longer supported.

What I am finding is that if I am passing less than 350 items into the array, I'm getting results:

```
SELECT count (*) FROM pgr_drivingDistance('
SELECT a.gid AS id ,a.source::INT AS source ,a.target::INT AS target ,a.length_m * 0.000621371::double precision AS cost FROM ways a WHERE a.x1> -121  AND a.x1 < -117 AND a.x2 >-121 AND a.x2<-117 AND a.y1<48 AND a.y1>44 AND a.y2<48 AND a.y2>44',
(SELECT array_agg(a.id) FROM ways_vertices_pgr a 
   INNER JOIN planet_osm_point b ON a.osm_id = b.osm_id 
   WHERE b.highway = 'motorway_junction' AND a.lat>44 AND a.lat<49.375 AND a.lon>-121 AND a.lon<-117 )
, 10, TRUE) a WHERE a.edge > - 1
 count
--------
 657531
(1 row)
```

And the following query returns no results

```
SELECT count (*) FROM pgr_drivingDistance('
SELECT a.gid AS id ,a.source::INT AS source ,a.target::INT AS target ,a.length_m * 0.000621371::double precision AS cost FROM ways a WHERE a.x1> -121  AND a.x1 < -117 AND a.x2 >-121 AND a.x2<-117 AND a.y1<48 AND a.y1>44 AND a.y2<48 AND a.y2>44',
(SELECT array_agg(a.id) FROM ways_vertices_pgr a 
  INNER JOIN planet_osm_point b ON a.osm_id = b.osm_id 
  WHERE b.highway = 'motorway_junction' AND a.lat>44 AND a.lat<49.377 AND a.lon>-121 AND a.lon<-117 )
, 10, TRUE) a WHERE a.edge > - 1
 count
--------
 0
(1 row)
```
And the only difference is that in the second query a.lat is restricted by 49.377, not 49.375, which pulls in one more item into the array. So now I have 350 items in the array instead of 349 and for some reason this causes pgr_drivingDistance to return no results at all. Is 349 instead of 350 a red herring?

**Comment by cvvergara** [comment]
@AlexBulankou 
This is the documentatin for v2.1:
http://docs.pgrouting.org/2.1/src/driving_distance/doc/dd_driving_distance_v3.html#pgr-driving-distance-v3
it does mention the array.

I need to recreate the problem with exactly the same data you are using.

**Comment by woodbri** [comment]
@AlexBulankou Can you create a simple example that reproduces this problem so we can reproduce this issue. You can use pgsql2shp to dump your edge table or ``pg_dump -Fp -f output.sql.gz -Z -t schema.tablename -x -O database -U postgres -h localhost `` or something closr to that and then attach the file to the ticket. If we can not reproduce it here it will be impossible to debug. Thanks.

**Comment by AlexBulankou** [comment]
@cvvergara @woodbri thanks! working on getting you the repro data you need.

**Comment by AlexBulankou** [comment]
I'm attaching the results of two queries in my repro, so you can use them to repro. 

`SELECT count (*) FROM pgr_drivingDistance('SELECT  id ,source , target , cost FROM attached ways.csv', (array from attached points.csv), 10, TRUE) a WHERE a.edge > - 1`

Right now the attached points.csv array has 350 items, so you should be able to repro (no results), but if you remove one of the items, it should return results. Please let me know if you have what you need.

[pgr_drivingDistance_repro.zip](https://github.com/pgRouting/pgrouting/files/157887/pgr_drivingDistance_repro.zip)

**Comment by cvvergara** [comment]
@AlexBulankou 
I will test as soon as I can, right now I am busy with the next release.
But maybe you can try with [V2.2-beta1](https://github.com/pgRouting/pgrouting/releases/tag/pgrouting-2.2.0-beta1) and see if it happens also there.

**Comment by AlexBulankou** [comment]
The issue seems to be that in the array of items, there's one item that is not connected to any edges in select query, so has nothing to do with the length of the array. Probably in the code that is combining the output, disconnected item causes early termination.

**Comment by cvvergara** [comment]
I made this simple test with the [sample data](http://docs.pgrouting.org/v2.2.0-beta1/doc/src/developer/sampledata.html#sampledata):
```
select * from pgr_drivingDistance('select * from edge_table', 1, 3);
 seq | node | edge | cost | agg_cost 
-----+------+------+------+----------
   1 |    1 |   -1 |    0 |        0
   2 |    2 |    1 |    1 |        1
   3 |    5 |    4 |    1 |        2
   4 |    6 |    8 |    1 |        3
   5 |    8 |    7 |    1 |        3
   6 |   10 |   10 |    1 |        3
(6 rows)

doc_data=# select * from pgr_drivingDistance('select * from edge_table', 5, 3);
 seq | node | edge | cost | agg_cost 
-----+------+------+------+----------
   1 |    5 |   -1 |    0 |        0
   2 |    2 |    4 |    1 |        1
   3 |    6 |    8 |    1 |        1
   4 |    8 |    7 |    1 |        1
   5 |   10 |   10 |    1 |        1
   6 |    1 |    1 |    1 |        2
   7 |    7 |    6 |    1 |        2
   8 |    9 |    9 |    1 |        2
   9 |   11 |   12 |    1 |        2
  10 |   13 |   14 |    1 |        2
  11 |    4 |   16 |    1 |        3
  12 |   12 |   15 |    1 |        3
(12 rows)

doc_data=# select * from pgr_drivingDistance('select * from edge_table', 25, 3);
 seq | node | edge | cost | agg_cost 
-----+------+------+------+----------
   1 |   25 |   -1 |    0 |        0
(1 row)

doc_data=# select * from pgr_drivingDistance('select * from edge_table', array[1,5,25], 3);
 seq | from_v | node | edge | cost | agg_cost 
-----+--------+------+------+------+----------
   1 |      1 |    1 |   -1 |    0 |        0
   2 |      1 |    2 |    1 |    1 |        1
   3 |      1 |    5 |    4 |    1 |        2
   4 |      1 |    6 |    8 |    1 |        3
   5 |      1 |    8 |    7 |    1 |        3
   6 |      1 |   10 |   10 |    1 |        3
   7 |      5 |    5 |   -1 |    0 |        0
   8 |      5 |    2 |    4 |    1 |        1
   9 |      5 |    6 |    8 |    1 |        1
  10 |      5 |    8 |    7 |    1 |        1
  11 |      5 |   10 |   10 |    1 |        1
  12 |      5 |    1 |    1 |    1 |        2
  13 |      5 |    7 |    6 |    1 |        2
  14 |      5 |    9 |    9 |    1 |        2
  15 |      5 |   11 |   12 |    1 |        2
  16 |      5 |   13 |   14 |    1 |        2
  17 |      5 |    4 |   16 |    1 |        3
  18 |      5 |   12 |   15 |    1 |        3
  19 |      5 |   25 |   -1 |    0 |        0
(19 rows)
```
The results miss the following record:
 select * from pgr_drivingDistance('select * from edge_table', 25, 3);
 seq | node | edge | cost | agg_cost 
-----+------+------+------+----------
   1 |   25 |   -1 |    0 |        0

**Comment by cvvergara** [comment]
Forgot to mention that test was done with develop branch:
```
doc_data=# select pgr_version();
                       pgr_version                       
---------------------------------------------------------
 (2.2.0,pgrouting-2.2.0-beta2dev,f82dff3,develop,1.54.0)
(1 row)
```

**Comment by cvvergara** [comment]
I am wrong about my comment, it does have the record, but the from_v value is wrong:
```
  19 |      5 |   25 |   -1 |    0 |        0
(19 rows)
```

Should be:
```
  19 |      25 |   25 |   -1 |    0 |        0
```

**Comment by cvvergara** [comment]
I think I've fixed it. (I just need to test with your data)
```
 select * from pgr_drivingDistance('select * from edge_table', array[1,5,25], 3);
 seq | from_v | node | edge | cost | agg_cost 
-----+--------+------+------+------+----------
   1 |      1 |    1 |   -1 |    0 |        0
   2 |      1 |    2 |    1 |    1 |        1
   3 |      1 |    5 |    4 |    1 |        2
   4 |      1 |    6 |    8 |    1 |        3
   5 |      1 |    8 |    7 |    1 |        3
   6 |      1 |   10 |   10 |    1 |        3
   7 |      5 |    5 |   -1 |    0 |        0
   8 |      5 |    2 |    4 |    1 |        1
   9 |      5 |    6 |    8 |    1 |        1
  10 |      5 |    8 |    7 |    1 |        1
  11 |      5 |   10 |   10 |    1 |        1
  12 |      5 |    1 |    1 |    1 |        2
  13 |      5 |    7 |    6 |    1 |        2
  14 |      5 |    9 |    9 |    1 |        2
  15 |      5 |   11 |   12 |    1 |        2
  16 |      5 |   13 |   14 |    1 |        2
  17 |      5 |    4 |   16 |    1 |        3
  18 |      5 |   12 |   15 |    1 |        3
  19 |     25 |   25 |   -1 |    0 |        0
(19 rows)
```

**Comment by cvvergara** [comment]
@AlexBulankou 
I finally loded the data (the file has some mistakes), I am following your instructions you give here: https://github.com/pgRouting/pgrouting/issues/519#issuecomment-192031152

First of all, this is what I am using:
```
SELECT pgr_version();
                       pgr_version                       
---------------------------------------------------------
 (2.2.0,pgrouting-2.2.0-beta2dev,f82dff3,develop,1.54.0)
(1 row)
```
Which is in develop branch right now. And for al I know it fixes your problem.
Please try using develop branch with your original queries.


I did a copy/paste of the query you that you mention  so that I reproduce the bug.
here is the query you wrote:
```
SELECT count (*) FROM pgr_drivingDistance(
         'SELECT id ,source , target , cost FROM attached ways.csv', 
         (array from attached points.csv),
          10, TRUE) a WHERE a.edge > - 1
```
This is what I get:
```
ERROR:  syntax error at or near "from"
LINE 1: ...e , target , cost FROM attached ways.csv', (array from attac...
```
So, I tried this query:
```
WITH setofid AS (SELECT array_agg(id) AS ids from points),
dd_results AS (SELECT * FROM
        pgr_driving Distance(
             'SELECT id ,source , target , cost FROM ways', (SELECT ids FROM setofid),
              10, TRUE) )
SELECT count(*) FROM dd_results WHERE edge > -1;
 count  
--------
 657531
(1 row)
```

So following your instructions I am going to remove a point form the query:
```
WITH setofid AS (SELECT array_agg(id) AS ids FROM points 
                   WHERE id NOT IN (SELECT id FROM points LIMIT 1)),
dd_results AS (SELECT * FROM
        pgr_drivingDistance(
             'SELECT id ,source , target , cost FROM ways', (SELECT ids FROM setofid), 
             10, TRUE) )
SELECT count(*) FROM dd_results where WHERE > -1;
 count  
--------
 656648
(1 row)
```

**Comment by cvvergara** [comment]
fixed in Tag pgrouting-2.2.0-RC1
